### PR TITLE
Refactor Discourse integration to ServiceObject classes

### PR DESCRIPTION
**Pull request by piotrekpasciak**
Monday:
https://aicrowd.monday.com/boards/363093650/pulses/429654890

Github Issue:
https://github.com/AIcrowd/AIcrowd/issues/1065

Discourse Network errors are handled in `discourse_api` gem:
https://github.com/discourse/discourse_api/blob/master/lib/discourse_api/client.rb#L143-L144

Change of uniqueness collision is 1 to (256^6 - Challenge.count), thus I've decided to ignore it.

### CODE REVIEW

File: app/services/discourse/create_category_service.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      response = create_category_request(ensure_uniqueness: retry_count.positive?)
+      challenge.update!(discourse_category_id: response['id'])
+
+      success
+    rescue DiscourseApi::UnprocessableEntity => e
+      raise e if retry_count.positive?
+
+      retry_count += 1
+
+      retry if e.message.include?('Category Name has already been taken')
```
Comment (skbly7) [inline app/services/discourse/create_category_service.rb]: That's nice to have random string only when needed! Well thought implementation. 💯

File: app/services/discourse/update_category_service.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
+module Discourse
+  class UpdateCategoryService < ::Discourse::BaseService
+    def initialize(challenge:)
+      @client    = prepare_api_client
+      @challenge = challenge
+    end
+
+    def call
```
Comment (skbly7) [inline app/services/discourse/update_category_service.rb]: (Minor) This function is similar to `create_cateogry_service -> call` and can be potentially refactored.

Some function say `def update_of_create` in Discourse::BaseService which knows to send create request v/s update based on parameters

File: spec/services/discourse/create_category_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
+require 'rails_helper'
```
Comment (skbly7) [inline spec/services/discourse/create_category_service_spec.rb]: 💯

File: app/services/discourse/update_category_service.rb (hunk 2 above)
Comment (piotrekpasciak) [reply app/services/discourse/update_category_service.rb]: That kind of action is usually called `upsert`. If we decide to implement it, I would simply merge both services into `UpsertService` class.

Comment (skbly7) [reply app/services/discourse/update_category_service.rb]: Sure. `UpsertService` would be good fit for this usecase.

**Pull request merged**